File: longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        HashSet<Character> set = new HashSet<>();
        int res = 0;
        int left = 0;
        int right = 0;

        while (right < s.length()) {
            if (!set.contains(s.charAt(right))) {
                res = Math.max(res, right - left + 1);
            } else {
                while (set.contains(s.charAt(right))) {
                    set.remove(s.charAt(left++));
                }
            }
            set.add(s.charAt(right++));
        }
        return res;
    }
}

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        map = {}
        left = 0
        longest = 0
        for right in range(len(s)):
            if s[right] not in map:
                map[s[right]] = 1
            else:
                map[s[right]] += 1
            while len(map) < right - left + 1:
                map[s[left]] -= 1
                if not map[s[left]]:
                    del map[s[left]]
                left += 1
            longest = max(longest, right - left + 1)
        return longest